rclpy/test/test_destruction.py

```diff
@@ -41,6 +44,28 @@ def test_destroy_node_twice():
         rclpy.shutdown(context=context)
 
 
+def test_destroy_node_while_spinning():
+    context = rclpy.context.Context()
+    rclpy.init(context=context)
+    try:
+        executor = rclpy.executors.SingleThreadedExecutor(context=context)
```

Comment:
since test is intended to check multi-threaded data access contention during shutdown, i think it would be nice to add `MultiThreadedExecutor` too.

Reply:
The specific issue that this test is supposed to show stems from destroying a node while another thread is spinning on that node. With knowledge of the code, I can tell you that it won't matter how many spinners there are - any spinning thread during a call to `destroy_node` will reproduce the issue.

That said, there isn't any reason not to use `MultiThreadedExecutor`, so I changed it in 649e399. Maybe it will uncover other issues.

Reply:
Actually, I was only partially correct. `SingleThreadedExecutor` and `MultiThreadedExecutor` behave differently when `context.shutdown()` is called. It appears that the `MultiThreadedExecutor` isn't propagating the `ExternalShutdownException` out of the call to `spin`.
The segfault is still happening when `rmw_cyclonedds_cpp` is used, but the exception on `rmw_fastrtps_cpp` gets lost.
Since that exception is actually something we want to see (indicating that the call to `context.shutdown()` is what terminated the spin, not the call to `node.destroy_node()`, I'm going to go back to `SingleThreadedExceutor`.